command/harness/dlite/execute.go

```diff
@@ -47,23 +55,63 @@ func (t *VMExecuteTask) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	ctxState().Add(cancel, req.ExecuteVMRequest.StageRuntimeID, task.ID)
-
 	req.ExecuteVMRequest.CorrelationID = task.ID
-	stepResp, err := harness.HandleStep(ctx, &req.ExecuteVMRequest, t.c.stageOwnerStore, &t.c.env, t.c.poolManager, t.c.metrics)
+	distributed := req.ExecuteVMRequest.Distributed
+	if distributed {
+		// create a temp token with step expiry + offset
+		token := ""
+		token, err = delegate.Token(audience, issuer, t.c.env.Dlite.AccountID, t.c.env.Dlite.AccountSecret, harness.StepTimeout+tokenExpiryOffset)
+		if err != nil {
+			logr.WithError(err).
+				WithField("stage_runtime_id", req.ExecuteVMRequest.StageRuntimeID).
+				Error("unable to generate token")
+			httphelper.WriteBadRequest(w, err)
+			return
+		}
+		// setting this only for distributed mode
+		req.ExecuteVMRequest.StartStepRequest.StageRuntimeID = req.ExecuteVMRequest.StageRuntimeID
+		req.ExecuteVMRequest.StepStatus = api.StepStatusConfig{
+			Endpoint:   t.c.env.Dlite.ManagerEndpoint,
+			AccountID:  t.c.env.Dlite.AccountID,
+			TaskID:     req.ExecuteVMRequest.TaskID,
+			DelegateID: task.DelegateInfo.ID,
+			Token:      token,
+		}
+	} else {
+		ctxState().Add(cancel, req.ExecuteVMRequest.StageRuntimeID, task.ID)
+	}
 
-	ctxState().DeleteTask(req.ExecuteVMRequest.StageRuntimeID, task.ID)
+	var stepResp *api.PollStepResponse
+	stepResp, err = harness.HandleStep(ctx, &req.ExecuteVMRequest, t.c.stageOwnerStore, &t.c.env, t.c.poolManager, t.c.metrics, distributed)
+	if !distributed {
+		ctxState().DeleteTask(req.ExecuteVMRequest.StageRuntimeID, task.ID)
+	}
 	if err != nil {
 		logr.WithError(err).
 			WithField("stage_runtime_id", req.ExecuteVMRequest.StageRuntimeID).
 			Error("could not execute step")
 		httphelper.WriteJSON(w, failedResponse(err.Error()), httpFailed)
 		return
 	}
-	resp := convert(stepResp)
-	resp.DelegateMetaInfo.HostName = t.c.delegateInfo.Host
-	resp.DelegateMetaInfo.ID = t.c.delegateInfo.ID
-	httphelper.WriteJSON(w, resp, httpOK)
+	t.handleResponse(w, stepResp, distributed)
+}
+
+func (t *VMExecuteTask) handleResponse(w http.ResponseWriter, resp *api.PollStepResponse, distributed bool) {
+	if distributed {
+		var r VMTaskExecutionResponse
+		if resp.Error == "" {
```

Comment:
From the code, it seems that we are just returning an empty poll response from `HandleStep` in the case of async right? Is this block of code needed?

Reply:
yes this is needed. Before poll we call the start step request as well. If some failure happens there then we need to handle that.